### PR TITLE
Improve monitor performance

### PR DESCRIPTION
**Pull request by Ana06**
This is improving both the old UI and the new one (Bootstrap), although it has a much bigger impact in the new one.

Solves https://github.com/openSUSE/open-build-service/issues/6554

Times with 74 packages and 40 repositories:

# NEW UI

### Before

14151ms (Views: 12719.2ms | ActiveRecord: 191.2ms | Backend: 726.9ms)

### Now

3750ms (Views: 2785.9ms | ActiveRecord: 19.9ms | Backend: 724.4ms)

# OLD UI

### Before

4685ms (Views: 3351.1ms | ActiveRecord: 176.3ms | Backend: 680.2ms)

### Now

3653ms (Views: 2681.3ms | ActiveRecord: 21.3ms | Backend: 715.8ms)

### CODE REVIEW

**Comment by Ana06** [comment]
[Improve Monitor Page Performance](https://trello.com/c/Pj3pWPd3/1036-improve-monitor-page-performance)

File: src/api/app/controllers/webui/project_controller.rb [hunk 1]
```diff
@@ -448,6 +448,8 @@ def monitor
         @repohash[repo].delete(arch) unless has_packages
       end
     end
+
+    @repohash = @repohash.sort
```
[the remark this answers is not in the document]
Comment (Ana06) [reply src/api/app/controllers/webui/project_controller.rb]: becase `sort!` doesn't exist for hashes...

File: src/api/app/views/webui/project/monitor.html.erb [hunk 2]
```diff
@@ -83,16 +83,16 @@
...
             <th colspan="<%= archlist.length %>"><span class="strong"><%= repo %></span></th>
           <% end -%>
         </tr>
         <!-- arch row -->
         <tr>
           <th></th>
-          <% @repohash.sort.each do |repo, archlist| -%>
-            <% archlist.sort.each do |arch| -%>
+          <% @repohash.each do |repo, archlist| -%>
+            <% archlist.sort!.each do |arch| -%>
```
[the remark this answers is not in the document]
Comment (Ana06) [reply src/api/app/views/webui/project/monitor.html.erb]: Doing it in the controller makes that this is not an improvement any more as I would have to iterate an extra time.

**Comment by Ana06** [comment]
@coolo 

> Your benchmark looks like cheating as you didn't change the backend calls, but picked runtimes where the backend changes dramatically :)

I put the benchmark just to show that this is not solving the other issue and that we go significatively down in SQL. It is anyway not accurate because depends on if there are other things running.

File: src/api/app/controllers/webui/project_controller.rb [hunk 3]
```diff
@@ -737,30 +738,24 @@ def monitor_set_filter(defaults)
...
-    @avail_arch_values = []
-    @avail_repo_values = []
-
-    @project.api_obj.repositories.each do |r|
-      @avail_repo_values << r.name
-      @avail_arch_values << r.architectures.pluck(:name)
-    end
-    @avail_arch_values = @avail_arch_values.flatten.uniq.sort!
-    @avail_repo_values = @avail_repo_values.flatten.uniq.sort!
+    repos = @project.api_obj.repositories
```
[the remark this answers is not in the document]
Comment (Ana06) [reply src/api/app/controllers/webui/project_controller.rb]: :confused:

[the remark this answers is not in the document]
Comment (coolo) [reply src/api/app/controllers/webui/project_controller.rb]: This is btw the only non-nitpick

File: src/api/app/controllers/webui/project_controller.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (coolo) [reply src/api/app/controllers/webui/project_controller.rb]: aehm, ok. that brings up another point: what you return here is no longer a repo __hash__

File: src/api/app/views/webui/project/monitor.html.erb (hunk 2 above)
[the remark this answers is not in the document]
Comment (coolo) [reply src/api/app/views/webui/project/monitor.html.erb]: well, no. package_buildresult iterates through the repos and all you would need to do is making sure the array in @repohash[repo] is sorted at that time. The repositories in @buildresult don't jump up and down, look at `osc api /build/home:coolo/_result`

File: src/api/app/controllers/webui/project_controller.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (Ana06) [reply src/api/app/controllers/webui/project_controller.rb]: It is a hash....

[the remark this answers is not in the document]
Comment (coolo) [reply src/api/app/controllers/webui/project_controller.rb]: a sorted hash is an array

File: src/api/app/controllers/webui/project_controller.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (Ana06) [reply src/api/app/controllers/webui/project_controller.rb]: Done: https://github.com/openSUSE/open-build-service/pull/6858

File: src/api/app/controllers/webui/project_controller.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (Ana06) [reply src/api/app/controllers/webui/project_controller.rb]: ah.. I didn't get what you mean... I can solve that :building_construction:

**Comment by coolo** [comment]
please rebase :)

File: src/api/app/controllers/webui/project_controller.rb [hunk 4]
```diff
@@ -448,6 +448,8 @@ def monitor
         @repohash[repo].delete(arch) unless has_packages
       end
     end
+
+    @repoarray = @repohash.sort
```
Comment (coolo) [inline src/api/app/controllers/webui/project_controller.rb]: You know what makes me happy? Telling you now that if repohash is not used in the views anymore, we should better not make it instance variable to make this clear :)

But I wonder: app/mixins/webui/load_buildresults.rb and spec/controllers/webui/project_controller_spec.rb access this variable. I guess this is obsolete

Comment (Ana06) [reply src/api/app/controllers/webui/project_controller.rb]: and we use it in different methods. We would have to return it in the method and there are other variables there we are not returning. I don't think this is that important. :thinking:

Comment (coolo) [reply src/api/app/controllers/webui/project_controller.rb]: You know how this works when you restart optimizing code, no? :)

getting into refactoring

Comment (coolo) [reply src/api/app/controllers/webui/project_controller.rb]: The problem is if you don't change the name, but don't change the users from expecting a hash either you get a problem just as well

Comment (Ana06) [reply src/api/app/controllers/webui/project_controller.rb]: it is changed! :wink:

**Comment by coolo** [comment]
Ana and I looked into it and the other users of repohash are different

**Comment by Ana06** [comment]
@coolo 

> Ana and I looked into it and the other users of repohash are different

what do you mean? :confused:

**Comment by coolo** [comment]
Referencing this:
> But I wonder: app/mixins/webui/load_buildresults.rb and spec/controllers/webui/project_controller_spec.rb
> access this variable. I guess this is obsolete

**Pull request merged**